Import transcript abundances from kallisto using tximport
================
Chelsea Herdman
February 5th, 2020

In order to perform differential expression analysis using DESeq2 on the
estimated transcript abundances produced by kallisto, we used the
tximport package in R.

We followed the vignette provided by Michael I. Love, Charlotte Soneson
and Mark D. Robinson found
[here](http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#salmon_with_inferential_replicates)

#### Prepare transcript to gene dataframe for gene-level summarization

Load required libraries.

``` r
library(tximport)
library(data.table)
library(ensembldb)
library(RMariaDB)
library(here)
```

Fetch gene annotations from Ensembl and create one to one transcript
id/gene id data frame.

``` r
txdb = makeTxDbFromEnsembl(organism="Danio rerio",
                           release=99,
                           circ_seqs=DEFAULT_CIRC_SEQS,
                           server="ensembldb.ensembl.org",
                           username="anonymous", password=NULL, port=0L,
                           tx_attrib=NULL)
```

    ## Fetch transcripts and genes from Ensembl ... OK
    ## Fetch exons and CDS from Ensembl ... OK
    ## Fetch chromosome names and lengths from Ensembl ...OK
    ## Gather the metadata ... OK
    ## Make the TxDb object ... OK

``` r
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- select(txdb, k, "GENEID", "TXNAME")
```

    ## 'select()' returned 1:1 mapping between keys and columns

``` r
head(tx2gene)
```

    ##               TXNAME             GENEID
    ## 1 ENSDART00000193350 ENSDARG00000114632
    ## 2 ENSDART00000181299 ENSDARG00000109882
    ## 3 ENSDART00000180768 ENSDARG00000111634
    ## 4 ENSDART00000184307 ENSDARG00000114112
    ## 5 ENSDART00000186443 ENSDARG00000116981
    ## 6 ENSDART00000189953 ENSDARG00000109637

``` r
length(unique(tx2gene$TXNAME))
```

    ## [1] 65905

``` r
length(unique(tx2gene$GENEID))
```

    ## [1] 37241

#### Prepare named vector of files

Use kallisto abundance.tsv files. These are listed in alphabetical order
in the kallisto\_out directory so ensure naming associates accurately
and that names associate with sample info table that will be used for
DESeq2.

``` r
file_path_vec = list.files(path=here("kallisto_quant_combined_z11"),
                   pattern="abundance.tsv",
                   recursive = TRUE,
                   full.names = TRUE)
all(file.exists(file_path_vec))
```

    ## [1] TRUE

``` r
file_tab = data.table(file_path=file_path_vec,
                      gnomex_id=basename(dirname(file_path_vec)))
file_tab[, sort_order:=as.integer(gsub("^14893X", "", gnomex_id))]
setorder(file_tab, sort_order)

files = file_tab$file_path
names(files) = file_tab$gnomex_id
```

#### Run tximport for gene level estimation

DESeq2 will be run both using the *original counts and offset* method
and the *bias corrected counts without an offset* method in order to
compare the two possibilities of correcting for transcript length.

##### Original counts and offset method

This txi object can be used directly with downstream DESeq2 analyses
using the DESeqDataSetFromTximport
function.

``` r
txi.kallisto.offset = tximport(files, type= "kallisto", tx2gene = tx2gene, 
                               ignoreTxVersion = TRUE)
```

    ## Note: importing `abundance.h5` is typically faster than `abundance.tsv`

    ## reading in files with read.delim (install 'readr' package for speed up)

    ## 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 
    ## summarizing abundance
    ## summarizing counts
    ## summarizing length

``` r
txi.kallisto.offset$counts[1:6, 1:6]
```

    ##                     14893X1   14893X2   14893X3  14893X4  14893X5   14893X6
    ## ENSDARG00000000001  374.706  591.0910  654.4320  515.936  659.246  382.5580
    ## ENSDARG00000000002 1199.290 1739.9100 2833.5700 2851.060 2885.750  958.4420
    ## ENSDARG00000000018 4476.000 5743.0000 4855.0000 2571.003 3121.000 2772.0000
    ## ENSDARG00000000019 3346.000 2919.0000 3523.0000 3638.000 3963.000 2973.0000
    ## ENSDARG00000000068 6639.606  904.0585  734.4076 1118.496 1898.591 5708.6379
    ## ENSDARG00000000069 1313.363 1175.5840 1344.0650 1233.627 1490.076  856.8193

``` r
txi.kallisto.offset$abundance[1:6, 1:6]
```

    ##                     14893X1   14893X2  14893X3   14893X4  14893X5  14893X6
    ## ENSDARG00000000001  1.59625  3.439700  3.33837  2.732100  2.75351  2.53349
    ## ENSDARG00000000002  4.83748  9.585880 13.68640 14.296100 11.41250  6.01083
    ## ENSDARG00000000018 15.68289 27.535900 20.41935 11.236120 10.74622 15.12662
    ## ENSDARG00000000019 12.68530 15.113800 15.99380 17.146800 14.73090 17.52730
    ## ENSDARG00000000068 36.03243  6.768312  4.84584  7.616929 10.36181 47.85675
    ## ENSDARG00000000069 18.40023 17.573100 19.39014 20.311761 18.24976 15.59607

##### Bias corrected counts without an offset

``` r
txi.kallisto.biascorr = tximport(files, type= "kallisto", tx2gene = tx2gene, 
                                 ignoreTxVersion = TRUE, 
                                 countsFromAbundance="lengthScaledTPM")
```

    ## Note: importing `abundance.h5` is typically faster than `abundance.tsv`

    ## reading in files with read.delim (install 'readr' package for speed up)

    ## 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 
    ## summarizing abundance
    ## summarizing counts
    ## summarizing length

``` r
txi.kallisto.biascorr$counts[1:6, 1:6]
```

    ##                      14893X1   14893X2   14893X3  14893X4  14893X5   14893X6
    ## ENSDARG00000000001  369.5747  587.5723  652.0484  511.428  650.705  381.8967
    ## ENSDARG00000000002 1182.7086 1729.1393 2822.8761 2825.938 2847.970  956.7940
    ## ENSDARG00000000018 4398.8430 5698.3753 4831.6812 2548.093 3076.552 2762.3561
    ## ENSDARG00000000019 3299.2331 2900.1826 3509.1984 3605.636 3910.548 2967.9234
    ## ENSDARG00000000068 6491.0692  899.5855  736.4373 1109.402 1905.260 5612.9515
    ## ENSDARG00000000069 1698.4615 1196.7962 1509.9319 1515.886 1719.434  937.2876

``` r
counts_tab = as.data.table(txi.kallisto.biascorr$counts)
counts_tab$ensembl_gene_id = rownames(txi.kallisto.biascorr$counts)
setcolorder(counts_tab, c("ensembl_gene_id", paste("14893X", 1:22, sep="")))
```

Save the counts table in order to use as input for DESeq2 (*see DESeq2
folder*)

``` r
fwrite(counts_tab, file=here("Tximport", "20200204_ribozero_counts_fromtximport_biascorrected.txt.gz"), sep="\t")
```
